### PR TITLE
Fixing bugs in backgroundMusic branch.

### PR DESCRIPTION
**Pull request by gupta2140**
Fixed a bug when music didn't play on restarting GCompris after shutting it down with Background music disabled.

### CODE REVIEW

**Comment by gupta2140** [comment]
The background music playlist repeats after all the music have been played.

File: src/core/main.qml [hunk 1]
```diff
@@ -108,6 +108,15 @@ Window {
         id: backgroundMusic
         muted: !ApplicationSettings.isBackgroundMusicEnabled
 
+        muteChangeHandler: function() {
+            if(muted) { volume = 0 }
```
Comment (petitlapin) [inline src/core/main.qml]: coding style (same for if below)

File: src/core/GCAudio.qml [hunk 2]
```diff
@@ -57,6 +57,18 @@ Item {
      */
     property alias source: audio.source
 
+    /**
+     * type: positive real number less than 1
+     * Determines instensity of the audio.
```
Comment (petitlapin) [inline src/core/GCAudio.qml]: intensity

**Comment by gupta2140** [comment]
Corrected coding style of "if" statements. Waiting for review!

**Comment by gupta2140** [comment]
Implemented pausing of background music and updated musical activities in which background music needs to be paused. Waiting for review!

File: src/core/main.qml [hunk 3]
```diff
@@ -315,12 +333,18 @@ Window {
...
+                            backgroundMusic.pause()
+                        }
                         return pushHTransition
                     }
                 } else {
                     if(properties.exitItem.isDialog) {
                         return popVTransition
                     } else {
+                        main.isBackgroundMusicEnabledInActivity = true
+                        backgroundMusic.resume()
```
Comment (petitlapin) [inline src/core/main.qml]: can it be handled via the on...Changed signal instead to avoid ?

File: src/core/GCAudio.qml [hunk 4]
```diff
@@ -88,11 +100,25 @@ Item {
...
+
+    //Pauses the currently playing audio
+    function pause() {
+        if(playbackState === Audio.PlayingState)
+            audio.pause()
+    }
+
+    //Resumes the current audio if it had been paused
+    function resume() {
+        if(playbackState === Audio.PausedState)
```
Comment (petitlapin) [inline src/core/GCAudio.qml]: I would start the music even if it is stopped

File: src/activities/explore_farm_animals/ExploreLevels.qml [hunk 5]
```diff
@@ -36,6 +36,8 @@ ActivityBase {
     property string url
     property bool hasAudioQuestions
 
+    isMusicalActivity: true
```
[the remark this answers is not in the document]
Comment (gupta2140) [reply src/activities/explore_farm_animals/ExploreLevels.qml]: done :+1:

File: src/activities/melody/Melody.qml [hunk 6]
```diff
@@ -59,7 +61,6 @@ ActivityBase {
 
         onStart: {
             items.backgroundMusicStatus = ApplicationSettings.isBackgroundMusicEnabled
```
[the remark this answers is not in the document]
Comment (gupta2140) [reply src/activities/melody/Melody.qml]: Removed this variable in the latest commit.

File: src/core/ActivityBase.qml [hunk 7]
```diff
@@ -121,6 +121,8 @@ Item {
      */
     property GCAudio backgroundMusic
 
+    property bool isMusicalActivity: false
```
[the remark this answers is not in the document]
Comment (gupta2140) [reply src/core/ActivityBase.qml]: Commented.

File: src/core/GCAudio.qml (hunk 4 above)
Comment (gupta2140) [reply src/core/GCAudio.qml]: Done :+1:

File: src/core/main.qml (hunk 3 above)
Comment (gupta2140) [reply src/core/main.qml]: onChanged signal of isBackgroundMusicEnabledInActivity?

Comment (gupta2140) [reply src/core/main.qml]: sure, done :)

**Comment by gupta2140** [comment]
All the above requested changes have been made. Waiting for review! :)

File: src/core/ActivityBase.qml (hunk 7 above)
[the remark this answers is not in the document]
Comment (petitlapin) [reply src/core/ActivityBase.qml]: it should explains what it does behind, knowing it's a musical activity is not useful by itself.

File: src/activities/melody/Melody.qml [hunk 8]
```diff
@@ -58,8 +60,6 @@ ActivityBase {
         }
 
         onStart: {
-            items.backgroundMusicStatus = ApplicationSettings.isBackgroundMusicEnabled
```
Comment (petitlapin) [inline src/activities/melody/Melody.qml]: shouldn't it be removed from the items object too?

Comment (gupta2140) [reply src/activities/melody/Melody.qml]: oops, sorry I didn't notice it. Removed now :+1:

File: src/core/ActivityBase.qml (hunk 7 above)
[the remark this answers is not in the document]
Comment (gupta2140) [reply src/core/ActivityBase.qml]: sure, done in the latest commit :)

**Comment by gupta2140** [comment]
Above suggested improvements are made. Waiting for review!

File: src/core/main.qml [hunk 9]
```diff
@@ -121,6 +158,8 @@ Window {
                 rccBackgroundMusic = ApplicationInfo.getBackgroundMusicFromRcc()
                 for(var i = 0; i < rccBackgroundMusic.length; i++)
```
Comment (petitlapin) [inline src/core/main.qml]: can you shuffle the music here before adding them to the list (Core.shuffle(rccBackgroundMusic))?

Comment (gupta2140) [reply src/core/main.qml]: yep, sure! I'll do it tomorrow (my exams ends tomorrow :D ) Thanks ;)

**Comment by gupta2140** [comment]
Committed here: https://cgit.kde.org/gcompris.git/commit/?h=backgroundMusic&id=f4ba64ed3594ad7073cbbc1fa3c596231240bf5f

**Pull request closed without merging**